File: backend/src/main/java/com/inventory/backend/service/ProductService.java
package com.inventory.backend.service;

import com.inventory.backend.model.Product;
import com.inventory.backend.model.User;
import com.inventory.backend.repository.ProductRepository;
import com.inventory.backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private UserRepository userRepository;

    private User getUserByUsername(String username) {
        return userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found: " + username));
    }

    public List<Product> getAllProducts(String username) {
        User user = getUserByUsername(username);
        return productRepository.findByUser(user);
    }

    public List<Product> searchProducts(String username, String name) {
        User user = getUserByUsername(username);
        return productRepository.findByUserAndNameContainingIgnoreCase(user, name);
    }

    public Optional<Product> getProductById(String username, Long id) {
        User user = getUserByUsername(username);
        return productRepository.findByIdAndUser(id, user);
    }

    public Product createProduct(String username, Product product) {
        User user = getUserByUsername(username);
        product.setUser(user);
        return productRepository.save(product);
    }

    public Product updateProduct(String username, Long id, Product productDetails) {
        User user = getUserByUsername(username);
        return productRepository.findByIdAndUser(id, user).map(product -> {
            product.setName(productDetails.getName());
            product.setPrice(productDetails.getPrice());
            product.setQuantity(productDetails.getQuantity());
            product.setCategory(productDetails.getCategory());
            return productRepository.save(product);
        }).orElseThrow(() -> new RuntimeException("Product not found with id " + id));
    }

    public void deleteProduct(String username, Long id) {
        User user = getUserByUsername(username);
        Product product = productRepository.findByIdAndUser(id, user)
                .orElseThrow(() -> new RuntimeException("Product not found with id " + id));
        productRepository.delete(product);
    }
}

File: frontend/src/services/product.service.js
import api from './api';

const getAllProducts = () => {
    return api.get('/products');
};

const getProductById = (id) => {
    return api.get(`/products/${id}`);
};

const createProduct = (product) => {
    return api.post('/products', product);
};

const updateProduct = (id, product) => {
    return api.put(`/products/${id}`, product);
};

const deleteProduct = (id) => {
    return api.delete(`/products/${id}`);
};

const searchProducts = (name) => {
    return api.get(`/products/search?name=${name}`);
};

const ProductService = {
    getAllProducts,
    getProductById,
    createProduct,
    updateProduct,
    deleteProduct,
    searchProducts
};

export default ProductService;
